App Crashes on iOS when navigating, after upgrading to Kotlin 1.9.20

## Crash on iOS When Navigating, after upgrade to Kotlin 1.9.20
After upgrading to Kotlin 1.9.20 
App is crashing in iOS, when it navigates
On Android, its working fine

### Stack Trace:
```
Fatal Exception: kotlin.ConcurrentModificationException
0  Medial                         0xdcc10 kfun:kotlin.collections.HashMap.Itr#checkForComodification(){} + 587 (HashMap.kt:587)
1  Medial                         0xdce70 kfun:kotlin.collections.HashMap.EntriesItr#next(){}kotlin.collections.HashMap.EntryRef<1:0,1:1> + 618 (HashMap.kt:618)
2  Medial                         0x911354 kfun:kotlin.collections.Iterator#next(){}1:0-trampoline + 18 (Iterator.kt:18)
3  Medial                         0x1177cc kfun:kotlin.sequences.FilteringSequence.object-1.calcNext#internal + 170 (Sequences.kt:170)
4  Medial                         0x11796c kfun:kotlin.sequences.FilteringSequence.object-1.hasNext#internal + 195 (Sequences.kt:195)
5  Medial                         0x637764 kfun:cafe.adriel.voyager.core.model.ScreenModelStore.onEach#internal + 1313 (_Sequences.kt:1313)
6  Medial                         0x74be54 kfun:cafe.adriel.voyager.navigator.Navigator#dispose(cafe.adriel.voyager.core.screen.Screen){} + 1 (ScreenLifecycleStore.kt:1)
7  Medial                         0x750eb8 kfun:cafe.adriel.voyager.navigator.internal.object-2.dispose#internal + 1855 (_Collections.kt:1855)
8  Medial                         0x1f2fe4 kfun:androidx.compose.runtime.DisposableEffectImpl.onForgotten#internal + 86 (Effects.kt:86)
9  Medial                         0x1edca4 kfun:androidx.compose.runtime.CompositionImpl.RememberEventDispatcher.dispatchRememberObservers#internal (Composition.kt)
10 Medial                         0x1ebdbc kfun:androidx.compose.runtime.CompositionImpl.applyChangesInLocked#internal + 828 (Composition.kt:828)
11 Medial                         0x1ec2ec kfun:androidx.compose.runtime.CompositionImpl#applyChanges(){} + 850 (Composition.kt:850)
12 Medial                         0x1fa6e8 kfun:androidx.compose.runtime.Recomposer.$runRecomposeAndApplyChanges$lambda$4$lambda$3$FUNCTION_REFERENCE$12.$<bridge-UNNB>invoke(kotlin.Long){}#internal (ListUtils.kt)
13 Medial                         0x90d828 kfun:kotlin.Function1#invoke(1:0){}1:1-trampoline + 1 ([K][Suspend]Functions:1)
14 Medial                         0x1d534c kfun:androidx.compose.runtime.BroadcastFrameClock#sendFrame(kotlin.Long){} + 42 (BroadcastFrameClock.kt:42)
15 Medial                         0x3b8580 kfun:androidx.compose.ui.window.ComposeWindow.object-2.render#internal + 512 (ComposeScene.skiko.kt:512)
16 Medial                         0x3ce164 kfun:androidx.compose.ui.window.SkikoUIView.object-3.render#internal + 93 (SkikoUIView.kt:93)
17 Medial                         0x3c45f8 kfun:androidx.compose.ui.window.MetalRedrawer.draw#internal + 341 (MetalRedrawer.kt:341)
18 Medial                         0x3c59d4 kfun:androidx.compose.ui.window.MetalRedrawer.$<init>$lambda$0$FUNCTION_REFERENCE$0.$<bridge-UNN>invoke(){}#internal + 253 (MetalRedrawer.kt:253)
19 Medial                         0x90dea8 kfun:kotlin.Function0#invoke(){}1:0-trampoline + 1 ([K][Suspend]Functions:1)
20 Medial                         0x3c7fa0 kfun:androidx.compose.ui.window.DisplayLinkProxy.$imp:handleDisplayLinkTick#internal + 454 (MetalRedrawer.kt:454)
21 QuartzCore                     0x29328 CA::Display::DisplayLink::dispatch_items(unsigned long long, unsigned long long, unsigned long long)
22 QuartzCore                     0x146e68 CA::Display::DisplayLink::dispatch_deferred_display_links(unsigned int)
23 UIKitCore                      0x64dd94 _UIUpdateSequenceRun
24 UIKitCore                      0xcb2894 schedulerStepScheduledMainSection
25 UIKitCore                      0xcb1df0 runloopSourceCallback
26 CoreFoundation                 0xd3128 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__
27 CoreFoundation                 0xdf7b4 __CFRunLoopDoSource0
28 CoreFoundation                 0x645e8 __CFRunLoopDoSources0
29 CoreFoundation                 0x7a0d4 __CFRunLoopRun
30 CoreFoundation                 0x7f3ec CFRunLoopRunSpecific
31 GraphicsServices               0x135c GSEventRunModal
32 UIKitCore                      0x39cf58 -[UIApplication _run]
33 UIKitCore                      0x39cbbc UIApplicationMain
34 SwiftUI                        0x1bcc50 OUTLINED_FUNCTION_895
35 SwiftUI                        0x1231ec block_copy_helper.1
36 SwiftUI                        0x10d294 OUTLINED_FUNCTION_901
37 Medial                         0x8340 main (iOSApp.swift)
38 ???                            0x1a9a00dec (Missing)
```

> Can you try this PR? #245
> 
> Just checkout to the branch and run `./gradlew -Pversion="1.0.0-SNAPSHOT" publishToMavenLocal`

Running this task fails
```
> Task :buildSrc:compileKotlin FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':buildSrc:compileKotlin'.
> Error while evaluating property 'compilerOptions.jvmTarget' of task ':buildSrc:compileKotlin'.
   > Failed to calculate the value of property 'jvmTarget'.
      > Unknown Kotlin JVM target: 21
```

I've even tried changing my java version, gradle JDK version to 21, nothing works

--- a/voyager-core/src/commonMain/kotlin/cafe/adriel/voyager/core/model/ScreenModelStore.kt
+++ b/voyager-core/src/commonMain/kotlin/cafe/adriel/voyager/core/model/ScreenModelStore.kt
@@ -113,7 +113,7 @@ public object ScreenModelStore : ScreenDisposable {
 
 
     private fun Map<String, *>.onEachHolder(holderKey: String, block: (String) -> Unit) =
-        asSequence()
+        toMap() // copy
             .filter { it.key.startsWith(holderKey) }
             .map { it.key }
             .forEach(block)

--- a/voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/ThreadSafeMap.native.kt
+++ b/voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/ThreadSafeMap.native.kt
@@ -1,5 +1,6 @@
 package cafe.adriel.voyager.core.concurrent
 
+import androidx.compose.runtime.toMutableStateList
 import kotlinx.atomicfu.locks.SynchronizedObject
 import kotlinx.atomicfu.locks.synchronized
 
@@ -29,11 +30,11 @@ public actual class ThreadSafeMap<K, V>(
     }
 
     override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
-        get() = synchronized(syncObject) { delegate.entries }
+        get() = synchronized(syncObject) { delegate.entries.toMutableSet() }
     override val keys: MutableSet<K>
-        get() = synchronized(syncObject) { delegate.keys }
+        get() = synchronized(syncObject) { delegate.keys.toMutableSet() }
     override val values: MutableCollection<V>
-        get() = synchronized(syncObject) { delegate.values }
+        get() = synchronized(syncObject) { delegate.values.toMutableList() }
 
     override fun clear() {
         synchronized(syncObject) { delegate.clear() }

--- a/voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/ThreadSafeMap.native.kt
+++ b/voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/ThreadSafeMap.native.kt
@@ -1,6 +1,5 @@
 package cafe.adriel.voyager.core.concurrent
 
-import androidx.compose.runtime.toMutableStateList
 import kotlinx.atomicfu.locks.SynchronizedObject
 import kotlinx.atomicfu.locks.synchronized
 

--- a/voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/ThreadSafeMap.native.kt
+++ b/voyager-core/src/nativeMain/kotlin/cafe.adriel.voyager.core/concurrent/ThreadSafeMap.native.kt
@@ -29,11 +29,11 @@ public actual class ThreadSafeMap<K, V>(
     }
 
     override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
-        get() = delegate.entries
+        get() = synchronized(syncObject) { delegate.entries }
     override val keys: MutableSet<K>
-        get() = delegate.keys
+        get() = synchronized(syncObject) { delegate.keys }
     override val values: MutableCollection<V>
-        get() = delegate.values
+        get() = synchronized(syncObject) { delegate.values }
 
     override fun clear() {
         synchronized(syncObject) { delegate.clear() }

